### PR TITLE
EPMDJ-2360 Redesign notification endpoints

### PR DESCRIPTION
**Pull request by ViktoriiaLebedeva**

### CODE REVIEW

File: core/src/main/kotlin/endpoints/openapi/DrillAdminEndpoints.kt [hunk 1]
```diff
@@ -182,59 +182,59 @@ class DrillAdminEndpoints(override val kodein: Kodein) : KodeinAware {
...
-                        val (notificationId) = notificationIdObject
-                        if (notificationsManager.read(notificationId)) {
-                            topicResolver.sendToAllSubscribed("/notifications")
-                            HttpStatusCode.OK to "Notification with id $notificationId successfully read"
-                        } else {
-                            HttpStatusCode.NotFound to ErrorResponse("Notification with id $notificationId not found")
-                        }
+                        HttpStatusCode.NotFound to ErrorResponse("Notification with id $notificationId not found")
                     }
                     logger.info { "Notification was reed with result: $response" }
```
Comment (LikhanovMaxim) [inline core/src/main/kotlin/endpoints/openapi/DrillAdminEndpoints.kt]: reed is a wrong word here)

File: core/src/main/kotlin/endpoints/openapi/DrillAdminEndpoints.kt [hunk 2]
```diff
@@ -182,59 +182,59 @@ class DrillAdminEndpoints(override val kodein: Kodein) : KodeinAware {
             }
 
             authenticate {
-                val notificationResponds = "Read Notification"
-                    .examples(
-                        example("notification", NotificationId("Some notification id"))
+                val notificationResponds = "Read all notifications"
+                    .responds(
+                        ok<Unit>(), notFound(), badRequest()
```
Comment (LikhanovMaxim) [inline core/src/main/kotlin/endpoints/openapi/DrillAdminEndpoints.kt]: notFound(), badRequest()
is it necessary?

File: core/src/main/kotlin/endpoints/openapi/DrillAdminEndpoints.kt [hunk 3]
```diff
@@ -182,59 +182,59 @@ class DrillAdminEndpoints(override val kodein: Kodein) : KodeinAware {
...
+                    logger.info { "Read notifications" }
+                    notificationsManager.readAll()
+                    call.respond(HttpStatusCode.OK, EmptyContent)
+                }
+            }
+
+            authenticate {
+                val notificationResponds = "Delete all notifications"
+                    .responds(
+                        ok<Unit>(), notFound()
```
Comment (LikhanovMaxim) [inline core/src/main/kotlin/endpoints/openapi/DrillAdminEndpoints.kt]: notFound()
is it necessary?

File: core/src/main/kotlin/router/Routers.kt [hunk 4]
```diff
@@ -86,14 +86,13 @@ object Routes {
             }
         }
 
-        //TODO rewrite notifications
         @Group(SYSTEM)
-        @Location("/notifications/read")
-        class ReadNotification
+        @Location("/notification")
+        class Notifications
```
Comment (LikhanovMaxim) [inline core/src/main/kotlin/router/Routers.kt]: Why did you change 'notification**S**' to 'notification'?

File: tests/src/test/kotlin/endpoints/DrillServerWsTest.kt [hunk 5]
```diff
@@ -116,36 +115,39 @@ internal class DrillServerWsTest {
...
+                handleHttpPatchRequest(locations.href(Routes.Api.Notification(firstNotificationId)), token)
                 currentNotifications = getCurrentNotifications(incoming, outgoing)
                 assertEquals(
                     NotificationStatus.READ,
                     currentNotifications.find { it.id == firstNotificationId }!!.status
                 )
                 assertNotificationsCounters(currentNotifications, 3, 2)
-                handleHttpPostRequest(locations.href(Routes.Api.DeleteNotification()), jsonId, token)
+                handleHttpDeleteRequest(locations.href(Routes.Api.Notification(firstNotificationId)), token)
                 getCurrentNotifications(incoming, outgoing)
```
Comment (LikhanovMaxim) [inline tests/src/test/kotlin/endpoints/DrillServerWsTest.kt]: Why do we invoke this method?

File: tests/src/test/kotlin/endpoints/DrillServerWsTest.kt [hunk 6]
```diff
@@ -116,36 +115,39 @@ internal class DrillServerWsTest {
...
                 val firstNotificationId = currentNotifications.first().id
-                val jsonId = NotificationId.serializer() stringify NotificationId(firstNotificationId)
-                handleHttpPostRequest(locations.href(Routes.Api.ReadNotification()), jsonId, token)
+                handleHttpPatchRequest(locations.href(Routes.Api.Notification(firstNotificationId)), token)
                 currentNotifications = getCurrentNotifications(incoming, outgoing)
                 assertEquals(
                     NotificationStatus.READ,
                     currentNotifications.find { it.id == firstNotificationId }!!.status
                 )
                 assertNotificationsCounters(currentNotifications, 3, 2)
```
Comment (LikhanovMaxim) [inline tests/src/test/kotlin/endpoints/DrillServerWsTest.kt]: Can we extract this type of block into methods?

**Review by LikhanovMaxim — CHANGES_REQUESTED** [review]
few comments

File: core/src/main/kotlin/notification/Manager.kt [hunk 7]
```diff
@@ -6,58 +6,53 @@ import com.epam.drill.admin.endpoints.agent.*
...
-        notifications[id] = readNotification
-        notifications[id]!!.id == readNotification.id
+    fun read(id: String): Boolean = notifications[id]?.let { notification ->
+        _notifications.update { it.put(id, notification.copy(read = true)) }
+        true
     } ?: false
 
     fun deleteAll() {
-        notifications.clear()
+        _notifications.update { it.minus(notifications.keys) }
```
Comment (viktorbll) [inline core/src/main/kotlin/notification/Manager.kt]: minus???

File: core/src/main/kotlin/endpoints/ServerWsTopics.kt [hunk 8]
```diff
@@ -92,8 +92,8 @@ class ServerWsTopics(override val kodein: Kodein) : KodeinAware {
                         ?.config?.let { json.parseJson(it) } ?: ""
                 }
 
-                topic<WsRoutes.GetNotifications> {
-                    notificationManager.allNotifications.sortedByDescending { it.date }
+                topic<WsNotifications> {
+                    notificationManager.notifications.values.reversed()
```
Comment (viktorbll) [inline core/src/main/kotlin/endpoints/ServerWsTopics.kt]: this is a hack, not a solution
???

File: core/src/main/kotlin/notification/Manager.kt [hunk 9]
```diff
@@ -6,58 +6,53 @@ import com.epam.drill.admin.endpoints.agent.*
...
+import java.util.*
 
-class NotificationManager(override val kodein: Kodein) : KodeinAware {
+internal val logger = KotlinLogging.logger {}
 
+class NotificationManager(override val kodein: Kodein) : KodeinAware {
     private val topicResolver: TopicResolver by instance()
     private val plugins: Plugins by instance()
     private val agentManager: AgentManager by instance()
+    private val _notifications = atomic(mapOf<String, Notification>().toPersistentMap())
```
Comment (viktorbll) [inline core/src/main/kotlin/notification/Manager.kt]: persistentMapOf

**Pull request merged**